module/zfs/spa.c

```diff
@@ -4650,6 +4805,145 @@ spa_ld_check_features(spa_t *spa, boolean_t *missing_feat_writep)
 	return (0);
 }
 
+struct load_chain_map_arg {
+	blkptr_t *bp;
+	spa_t *spa;
+	uint64_t *error;
+};
+
+static int
+load_chain_map_claim_blk_cb(spa_t *spa, const blkptr_t *bp, void *arg)
+{
+	(void) arg;
+	int error = metaslab_claim(spa, bp,
+	    spa_get_dsl(spa)->dp_tx.tx_open_txg);
+	if (error == ENOENT)
+		error = 0;
+	return (error);
+}
+
+static int
+load_chain_map_claim_lr_cb(spa_t *spa, const lr_t *lrc, void *arg)
+{
+	(void) arg;
+	lr_write_t *lr = (lr_write_t *)lrc;
+	blkptr_t *bp = &lr->lr_blkptr;
+	if (lrc->lrc_txtype != TX_WRITE || BP_IS_HOLE(bp))
+		return (0);
+	int error = metaslab_claim(spa, bp,
+	    spa_get_dsl(spa)->dp_tx.tx_open_txg);
```

Comment:
`TX_WRITE` and `TX_CLONE_RANGE` expected to have a block pointers from the child pool, where actual data are stored.  Meanwhile it seems you are passing spa of the log pool here.  Am I wrong?

Reply:
TX_WRITE does not have block pointers from the child pool; there are a few different types of write records in the ZIL.  Of them, WR_INDIRECT writes would, but they are not done if there is a slog (including a shared log) according to `zil.h`. Looking at the code for zfs_log_write and zvol_log_write, that may not be quite right, and might need some tweaking.

Reply:
> Of them, WR_INDIRECT writes would, but they are not done if there is a slog

Right. But now it is just a question of performance optimization which might be changed, while you turn it into a hard requirement.